### ISSUE TITLE
Bug in "Don't use STRREF for pointer diff in string.find()." commit

### ISSUE THREAD

**Issue by sonyps5201314** [LuaJIT/LuaJIT#540]
follow is my code snippet:
```lua
function GameFrame:getBodyByName( param )
	
	if #self.m_bodyList ~= 0 then
		for i=1,#self.m_bodyList do
			local sublist = self.m_bodyList[i]
			local k = sublist.k
	
			if k == param then
				local points = sublist.p
				local physicsBody = cc.PhysicsBody:create(PHYSICS_INFINITY, PHYSICS_INFINITY)
				for s=1,#points do
					local onePoint = points[s]
					local resultPoints = {}
					for t=1,#onePoint do
						local vector = onePoint[t]
						--去掉大括号
						local result = string.sub(vector, 2, -2)
						local len = string.len(result)
						local dindex = string.find(result,",")
						print("vector=" .. vector .. " result=" .. result .. " dindex=" .. dindex)

						local subx = string.sub(result,1,dindex-1)
						local x = tonumber(subx)
						local suby = string.sub(result,dindex+1,len)
						local y = tonumber(suby)
		
						local p = cc.p(x,y)
						table.insert(resultPoints, p)
					end
				

					local center = cc.PhysicsShape:getPolyonCenter(resultPoints)
				    local shape = cc.PhysicsShapePolygon:create(resultPoints,cc.PHYSICSBODY_MATERIAL_DEFAULT,cc.p(-center.x, -center.y))
				    physicsBody:addShape(shape)
					physicsBody:setGravityEnable(false)
					return physicsBody

				end
			 break
			end
		end

	end

end
```
follow is the run console rusult:
```
setBlendFunc of cc.Sprite will deprecate two int parameter form,please pass a ta
ble like {src = xx, dst = xx} as a parameter
warning: don't update it again
cocos2d: SpriteFrameCache: Frame '' isn't found
warning: don't update it again
cocos2d: SpriteFrameCache: Frame '' isn't found
warning: don't update it again
warning: don't update it again
warning: don't update it again
warning: don't update it again
warning: don't update it again
warning: don't update it again
warning: don't update it again
warning: don't update it again
warning: don't update it again
warning: don't update it again
warning: don't update it again
warning: don't update it again
warning: don't update it again
warning: don't update it again
warning: don't update it again
warning: don't update it again
warning: don't update it again
cocos2d: removeChildByTag(tag = 1000): child not found!
cocos2d: removeChildByTag(tag = 205): child not found!
cocos2d: removeChildByTag(tag = 100): child not found!
cocos2d: removeChildByTag(tag = 71): child not found!
cocos2d: removeChildByTag(tag = 100): child not found!
cocos2d: removeChildByTag(tag = 72): child not found!
cocos2d: removeChildByTag(tag = 100): child not found!
cocos2d: removeChildByTag(tag = 73): child not found!
cocos2d: removeChildByTag(tag = 100): child not found!
cocos2d: removeChildByTag(tag = 74): child not found!
cocos2d: removeChildByTag(tag = 100): child not found!
cocos2d: removeChildByTag(tag = 76): child not found!
cocos2d: removeChildByTag(tag = 100): child not found!
cocos2d: removeChildByTag(tag = 77): child not found!
cocos2d: removeChildByTag(tag = 100): child not found!
cocos2d: removeChildByTag(tag = 78): child not found!
cocos2d: removeChildByTag(tag = 75): child not found!
cocos2d: removeChildByTag(tag = 1): child not found!
cocos2d: removeChildByTag(tag = 1): child not found!
cocos2d: removeChildByTag(tag = 4): child not found!
cocos2d: removeChildByTag(tag = 5): child not found!
cocos2d: removeChildByTag(tag = 3): child not found!
[LUA-print]
**********
ccp(x,y) was deprecated please use cc.p(x,y) instead.
**********
[LUA-print] vector={ 68.00000,-26.50000 } result= 68.00000,-26.50000  dindex=10
[LUA-print] vector={ 59.00000,-72.50000 } result= 59.00000,-72.50000  dindex=10
[LUA-print] vector={ -51.00000,-68.50000 } result= -51.00000,-68.50000  dindex=1
1
[LUA-print] vector={ -66.00000,-25.50000 } result= -66.00000,-25.50000  dindex=1
1
[LUA-print] vector={ 5.00000,43.50000 } result= 5.00000,43.50000  dindex=9
[LUA-print]
**********
cc.Node:setZOrder was deprecated please use cc.Node:setLocalZOrder instead.
**********
[LUA-print] vector={ -3.00000,-24.00000 } result= -3.00000,-24.00000  dindex=10
[LUA-print] vector={ -6.00000,26.00000 } result= -6.00000,26.00000  dindex=10
[LUA-print] vector={ 8.00000,17.00000 } result= 8.00000,17.00000  dindex=9
[LUA-print]
**********
cc.Node:setZOrder was deprecated please use cc.Node:setLocalZOrder instead.
**********
[LUA-print] vector={ 29.50000,-1.50000 } result= 29.50000,-1.50000  dindex=10
[LUA-print] vector={ 4.50000,-41.50000 } result= 4.50000,-41.50000  dindex=9
[LUA-print] vector={ -28.50000,4.50000 } result= -28.50000,4.50000  dindex=11
[LUA-print] vector={ 0.50000,46.50000 } result= 0.50000,46.50000  dindex=9
[LUA-print]
**********
cc.Node:setZOrder was deprecated please use cc.Node:setLocalZOrder instead.
**********
[LUA-print] vector={ 14.50000,45.50000 } result= 14.50000,45.50000  dindex=10
[LUA-print] vector={ 0.50000,-39.50000 } result= 0.50000,-39.50000  dindex=9
[LUA-print] vector={ -24.50000,1.50000 } result= -24.50000,1.50000  dindex=11
[LUA-print] vector={ -24.50000,44.50000 } result= -24.50000,44.50000  dindex=11
[LUA-print]
**********
cc.Node:setZOrder was deprecated please use cc.Node:setLocalZOrder instead.
**********
[LUA-print] vector={ 17.00000,36.50000 } result= 17.00000,36.50000  dindex=10
[LUA-print] vector={ 3.00000,-44.50000 } result= 3.00000,-44.50000  dindex=9
[LUA-print] vector={ -16.00000,11.50000 } result= -16.00000,11.50000  dindex=11
[LUA-print] vector={ -6.00000,43.50000 } result= -6.00000,43.50000  dindex=10
[LUA-print]
**********
cc.Node:setZOrder was deprecated please use cc.Node:setLocalZOrder instead.
**********
[LUA-print] vector={ 15.50000,25.50000 } result= 15.50000,25.50000  dindex=10
[LUA-print] vector={ 1.50000,-33.50000 } result= 1.50000,-33.50000  dindex=9
[LUA-print] vector={ -20.50000,-8.50000 } result= -20.50000,-8.50000  dindex=11
[LUA-print] vector={ -12.50000,25.50000 } result= -12.50000,25.50000  dindex=11
[LUA-print]
**********
cc.Node:setZOrder was deprecated please use cc.Node:setLocalZOrder instead.
**********
[LUA-print] vector={ 17.00000,36.50000 } result= 17.00000,36.50000  dindex=10
[LUA-print] vector={ 3.00000,-44.50000 } result= 3.00000,-44.50000  dindex=9
[LUA-print] vector={ -16.00000,11.50000 } result= -16.00000,11.50000  dindex=11
[LUA-print] vector={ -6.00000,43.50000 } result= -6.00000,43.50000  dindex=10
[LUA-print]
**********
cc.Node:setZOrder was deprecated please use cc.Node:setLocalZOrder instead.
**********
[LUA-print] vector={ 12.00000,-17.50000 } result= 12.00000,-17.50000  dindex=10
[LUA-print] vector={ -15.00000,-40.50000 } result= -15.00000,-40.50000  dindex=1
1
[LUA-print] vector={ -17.00000,19.50000 } result= -17.00000,19.50000  dindex=11
[LUA-print] vector={ 13.00000,50.50000 } result= 13.00000,50.50000  dindex=10
[LUA-print]
**********
cc.Node:setZOrder was deprecated please use cc.Node:setLocalZOrder instead.
**********
[LUA-print] vector={ -3.00000,-24.00000 } result= -3.00000,-24.00000  dindex=10
[LUA-print] vector={ -6.00000,26.00000 } result= -6.00000,26.00000  dindex=10
[LUA-print] vector={ 8.00000,17.00000 } result= 8.00000,17.00000  dindex=9
[LUA-print]
**********
cc.Node:setZOrder was deprecated please use cc.Node:setLocalZOrder instead.
**********
[LUA-print] vector={ -39.00000,-43.00000 } result= -39.00000,-43.00000  dindex=1
1
[LUA-print] vector={ -29.00000,46.00000 } result= -29.00000,46.00000  dindex=11
[LUA-print] vector={ 28.00000,46.00000 } result= 28.00000,46.00000  dindex=10
[LUA-print] vector={ 40.00000,-44.00000 } result= 40.00000,-44.00000  dindex=10
[LUA-print]
**********
cc.Node:setZOrder was deprecated please use cc.Node:setLocalZOrder instead.
**********
[LUA-print] vector={ -39.00000,-43.00000 } result= -39.00000,-43.00000  dindex=1
1
[LUA-print] vector={ -29.00000,46.00000 } result= -29.00000,46.00000  dindex=11
[LUA-print] vector={ 28.00000,46.00000 } result= 28.00000,46.00000  dindex=10
[LUA-print] vector={ 40.00000,-44.00000 } result= 40.00000,-44.00000  dindex=10
[LUA-print]
**********
cc.Node:setZOrder was deprecated please use cc.Node:setLocalZOrder instead.
**********
[LUA-print] vector={ 44.00000,111.00000 } result= 44.00000,111.00000  dindex=10
[LUA-print] vector={ -19.00000,-105.00000 } result= -19.00000,-105.00000  dindex
=11
[LUA-print] vector={ -34.00000,110.00000 } result= -34.00000,110.00000  dindex=1
1
[LUA-print]
**********
cc.Node:setZOrder was deprecated please use cc.Node:setLocalZOrder instead.
**********
[LUA-print] GameFrameEngine:onSocketUserEnter 已有自己信息，不再次处理自己信息
[LUA-print] vector={ 15.50000,25.50000 } result= 15.50000,25.50000  dindex=10
[LUA-print] vector={ 1.50000,-33.50000 } result= 1.50000,-33.50000  dindex=9
[LUA-print] vector={ -20.50000,-8.50000 } result= -20.50000,-8.50000  dindex=11
[LUA-print] vector={ -12.50000,25.50000 } result= -12.50000,25.50000  dindex=11
[LUA-print]
**********
cc.Node:setZOrder was deprecated please use cc.Node:setLocalZOrder instead.
**********
[LUA-print] vector={ 14.50000,45.50000 } result= 14.50000,45.50000  dindex=10
[LUA-print] vector={ 0.50000,-39.50000 } result= 0.50000,-39.50000  dindex=9
[LUA-print] vector={ -24.50000,1.50000 } result= -24.50000,1.50000  dindex=11
[LUA-print] vector={ -24.50000,44.50000 } result= -24.50000,44.50000  dindex=-33
6426580
[LUA-print] [string ".\cocos/cocos2d/Cocos2d.lua"]:25: attempt to index local '_
x' (a nil value)
stack traceback:
        [string ".\game/buyu/fishlk/src/models/GameFrame.lua"]:123: in function
'getBodyByType'
        [string ".\game/buyu/fishlk/src/views/layer/Fish1.lua"]:724: in function
 'initPhysicsBody'
        [string ".\game/buyu/fishlk/src/views/layer/Fish1.lua"]:502: in function
 'initAnim'
        [string ".\game/buyu/fishlk/src/views/GameLayer.lua"]:1905: in function
'onSubCreateMoreFishd'
        [string ".\game/buyu/fishlk/src/views/GameLayer.lua"]:1092: in function
'onEventGameMessage'
        [string ".\client/src/plaza/models/GameFrameEngine.lua"]:117: in functio
n 'onSocketEvent'
        [string ".\client/src/plaza/models/BaseFrame.lua"]:147: in function 'onS
ocketCallBack'
        [string ".\client/src/plaza/models/BaseFrame.lua"]:103: in function <[st
ring ".\client/src/plaza/models/BaseFrame.lua"]:102>
```
as you see,'local dindex = string.find(result,",")' will get a wrong result,after several calls. in this instance,'dindex' is -336426580 but it correct result should is 11.
Please correct it,thanks.

**Comment by mraleph** [comment]
Yeah, I don't think the fix that @MikePall landed was correct: the patch is missing the fact that `trsptr` defined as `emitir(IRT(IR_STRREF, IRT_PGC), trstr, trstart)` does not actually have to point into `trstr` at all - because `STRREF SNEW` can be folded in such a way that `STRREF` would be pointing into entirely different string.  

That is why I think you are bound to do arithmetic with `trsptr` rather than `trstr`.

**Comment by mraleph** [comment]
Reduced reproduction can be

```lua
jit.opt.start("hotloop=1")
local value = "____________________abc____________________"
for i = 1, 20 do
  value = string.sub(value, 2, -2)
  local pos_b = string.find(value, "b", 2, true)
  assert(pos_b == (22 - i), "FAIL: position of 'b' is " .. pos_b)
end

print("OK")
```

**Comment by MikePall** [comment]
Corrected the fix in #505. Thanks @mraleph!
